File: articles/cell-structure-membranes-organelles/c/cell_architecture_kernel.c
/*
 * Compact cell-architecture kernel in C.
 */

#include <math.h>
#include <stdio.h>

double sphere_surface_area(double radius) {
    return 4.0 * M_PI * radius * radius;
}

double sphere_volume(double radius) {
    return (4.0 / 3.0) * M_PI * radius * radius * radius;
}

double surface_area_to_volume(double radius) {
    return sphere_surface_area(radius) / sphere_volume(radius);
}

double permeability_flux(double permeability, double c_out, double c_in) {
    return permeability * (c_out - c_in);
}

double diffusive_flux(double diffusion_coefficient, double gradient) {
    return -diffusion_coefficient * gradient;
}

double organelle_fraction(double organelle_area, double cell_area) {
    return organelle_area / cell_area;
}

double organelle_density(double count, double cell_area) {
    return count / cell_area;
}

int main(void) {
    double radius = 5.0;

    printf("surface_area_um2=%.6f\n", sphere_surface_area(radius));
    printf("volume_um3=%.6f\n", sphere_volume(radius));
    printf("sa_to_volume=%.6f\n", surface_area_to_volume(radius));
    printf("permeability_flux=%.6f\n", permeability_flux(0.05, 10.0, 3.0));
    printf("diffusive_flux=%.6f\n", diffusive_flux(2.0, -0.8));
    printf("mitochondrial_fraction=%.6f\n", organelle_fraction(62.0, 420.0));
    printf("lysosome_density=%.6f\n", organelle_density(18.0, 420.0));

    return 0;
}

File: articles/cell-structure-membranes-organelles/fortran/cell_architecture_kernel.f90
! Compact cell-architecture numerical kernel in Fortran.

program cell_architecture_kernel
  implicit none

  real :: radius
  real :: surface_area
  real :: volume
  real :: sa_to_volume
  real :: permeability
  real :: c_out
  real :: c_in
  real :: membrane_flux
  real :: diffusion_coefficient
  real :: gradient
  real :: diff_flux
  real :: cell_area
  real :: mitochondrial_area
  real :: lysosome_count
  real :: mito_fraction
  real :: lysosome_density

  radius = 5.0
  surface_area = 4.0 * 3.14159265 * radius ** 2
  volume = (4.0 / 3.0) * 3.14159265 * radius ** 3
  sa_to_volume = surface_area / volume

  permeability = 0.05
  c_out = 10.0
  c_in = 3.0
  membrane_flux = permeability * (c_out - c_in)

  diffusion_coefficient = 2.0
  gradient = -0.8
  diff_flux = -diffusion_coefficient * gradient

  cell_area = 420.0
  mitochondrial_area = 62.0
  lysosome_count = 18.0
  mito_fraction = mitochondrial_area / cell_area
  lysosome_density = lysosome_count / cell_area

  print *, "Surface area:", surface_area
  print *, "Volume:", volume
  print *, "SA:V:", sa_to_volume
  print *, "Membrane flux:", membrane_flux
  print *, "Diffusive flux:", diff_flux
  print *, "Mitochondrial fraction:", mito_fraction
  print *, "Lysosome density:", lysosome_density
end program cell_architecture_kernel
